### PR TITLE
example: add and document dotfiles usage

### LINKED ISSUE #1765: example: example template with dotfiles usage
## What is your suggestion?

Add a template to `examples/` which leverages dotfiles as a developer parameter.

## Why do you want this feature?

Show use of how `coder dotfiles` can be used with templates

## Are there any workarounds to get this functionality today?

Run `coder dotfiles` once you SSH into a workspace 

## Are you interested in submitting a PR for this?

Yes

### PR DESCRIPTION
**Pull request by bpmct**
Closes #1765 #1274.

### CODE REVIEW

File: examples/templates/docker-with-dotfiles/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  see https://dotfiles.github.io
+  EOF
+    # The codercom/enterprise-* images are only built for amd64
+  default = ""
+}
+
+resource "coder_agent" "dev" {
+  ...
+  startup_script = var.dotfiles_uri != "" ? "/tmp/tmp.coder*/coder dotfiles -y ${var.dotfiles_uri}" : null
```
Comment (bpmct) [inline examples/templates/docker-with-dotfiles/README.md:32]: While `coder` was in PATH, it didn't seem to work in the startup script. Even when I specified /bin/bash. Any ideas on how to use the agent's shell?

I'm also on a slightly older build. I'll try upgrading and see if this fixes it

File: examples/templates/docker-with-dotfiles/main.tf [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  see https://dotfiles.github.io
+  EOF
+    # The codercom/enterprise-* images are only built for amd64
+  default = ""
+}
+
+resource "coder_agent" "dev" {
+  arch = var.step2_arch
+  os   = "linux"
+  startup_script = var.dotfiles_uri != "" ? "/tmp/tmp.coder*/coder dotfiles -y ${var.dotfiles_uri}" : null
```
Comment (f0ssel) [inline examples/templates/docker-with-dotfiles/main.tf:91]: `coder` is in the path now, no need to do this.

**Review by f0ssel — APPROVED** [review]
lgtm. Is there any way to simplify the template? Ideally we just want to highlight the usage of the startup_script parameter. But if not no biggie.

File: examples/templates/docker-with-dotfiles/main.tf [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  You can specify a different host in the template file and
+  surpress this warning.
+  EOF
+  validation {
+    condition     = contains(["Continue using /var/run/docker.sock on the Coder host"], var.step1_docker_host_warning)
+    error_message = "Cancelling template create."
+  }
+
+  sensitive = true
+}
```
Comment (kylecarbs) [inline examples/templates/docker-with-dotfiles/main.tf:32]: The Docker provider will use `DOCKER_HOST` by default, so we should be able to just exclude it from the provider and have this _just work_.

File: examples/templates/docker-with-dotfiles/main.tf (hunk 2 above)
Comment (bpmct) [reply examples/templates/docker-with-dotfiles/main.tf:91]: While coder was in PATH, it didn't seem to work in the startup script. Even when I specified /bin/bash. Any ideas on how to use the agent's shell?

I'm also on a slightly older build. I'll try upgrading and see if this fixes it

File: examples/templates/docker-with-dotfiles/main.tf (hunk 3 above)
Comment (bpmct) [reply examples/templates/docker-with-dotfiles/main.tf:32]: Sadly, many systems do not have `DOCKER_HOST` set

Comment (bpmct) [reply examples/templates/docker-with-dotfiles/main.tf:32]: However, I removed this warning since this just demonstrates dotfiles usage

**Comment by bpmct** [comment]
@f0ssel - I've simplified the template, I agree we should just focus on dotfiles usage in the example.

However, two issues, which I think are related:

1) the ssh key doesn't seem to be present in the machine at the time of startup_script
2) the Coder binary is not present at the time of startup_script
   ```sh
    /bin/bash: coder: No such file or directory
    coder: No such file or directory
    ```

However, when I ssh into the workspace, `coder` works. Any ideas?

**Comment by f0ssel** [comment]
@kylecarbs could you explain why this is the case? Are we running startup script via some other way than through `coder agent`?

File: examples/templates/docker-with-dotfiles/README.md (hunk 1 above)
Comment (mafredri) [reply examples/templates/docker-with-dotfiles/README.md:32]: The problem is that the PATH is not set _for_ the agent.

One option is to set the `PATH` earlier, before agent start, e.g. here: https://github.com/coder/coder/blob/d0ed107b08f83d7b5c1f4794937e16adda0137b1/provisionersdk/agent.go#L35

Something like `export PATH=${PATH}:${BINARY_LOCATION}`.

**OR** we can add it to the path for the startup script here:

https://github.com/coder/coder/blob/b65259f95e53e96cd8daf21a5c3de46e51a56f85/agent/agent.go#L176

Like we do here:

https://github.com/coder/coder/blob/b65259f95e53e96cd8daf21a5c3de46e51a56f85/agent/agent.go#L351

File: examples/templates/docker-with-dotfiles/main.tf (hunk 2 above)
Comment (bpmct) [reply examples/templates/docker-with-dotfiles/main.tf]: Resolved in #2099

**Pull request merged**